### DISCUSSION TITLE
Union type str | bytes where only one is possible

### DISCUSSION THREAD

**Discussion by leifwar** [benbenbang/types-confluent-kafka#216] [Q&A]
Several of the functions are typed with the union `str | bytes | None` where it should be either `bytes | None` for Python 3 and `str | None` for Python 2 (given by `cfl_PyBin`). Anyone using these stubs would then have to deal with a type hint that can never happen. Would it be possible to either drop the Python 2 type hints, since it should not be used anymore (https://www.python.org/doc/sunset-python-2/), or provide separate stubs for the two python versions?

**Comment by benbenbang** [comment]
> Several of the functions are typed with the union `str | bytes | None` where it should be either `bytes | None` for Python 3 and `str | None` for Python 2 (given by `cfl_PyBin`). Anyone using these stubs would then have to deal with a type hint that can never happen. Would it be possible to either drop the Python 2 type hints, since it should not be used anymore (https://www.python.org/doc/sunset-python-2/), or provide separate stubs for the two python versions?


Hello @leifwar,

Thank you for the feedback.

Yes, some versions are sunset, even for certain Python 3 versions; they should not be used anymore. This is a very good point, and I am aware of that; however, the issue is that mypy and pyright are not aware of this. Since some versions of Python do not support function signature overloading as of today, I have no choice but to retain all the signatures that are originally from the Confluent Kafka Python package.

Welcome for all the feedback 🙏🏻

Thanks again.

**Reply by leifwar** [reply]
I see your point and it makes sense if you're implementing something that supports both Python 2 and Python 3.

**Comment by anti-social** [comment]
Hi there.

> I have no choice but to retain all the signatures that are originally from the Confluent Kafka Python package.

@benbenbang Could you clarify your statement? Why a typing library should support Python 2 if it does not support type annotations at all?

**Reply by anti-social** [reply]
Yeah, it is possible to use type hints in Python 2 in comments. But what I meant is that the library is definitelty Python 3 only.

**Reply by benbenbang** [reply]
True, it's Python 3 only. 

When I implemented this, I just thought that I needed to follow the original implementation. That's why you see that.
